File: java/L600/LeetCode0630.java
package L600;

import java.util.Arrays;
import java.util.PriorityQueue;

public class LeetCode0630 {
    public int scheduleCourse(int[][] courses) {
        Arrays.sort(courses, (a, b) -> a[1] - b[1]);
        PriorityQueue<Integer> pq = new PriorityQueue<>((a, b) -> b - a);
        int sum = 0;
        for (int[] course : courses) {
            if (sum + course[0] <= course[1]) {
                sum += course[0];
                pq.offer(course[0]);
            } else {
                sum += course[0];
                pq.offer(course[0]);
                sum -= pq.poll();
            }
        }
        return pq.size();
    }
}

File: cpp/L600/LeetCode0630.cc
#include <algorithm>
#include <vector>
using std::vector;
#include <queue>

class Solution {
public:
    int scheduleCourse(vector<vector<int>>& courses) {
        std::sort(courses.begin(), courses.end(), [](vector<int> const &c1, vector<int> const &c2) {
            return c1[1] < c2[1];
        });
        auto f{[&](int x, int y) {
            return courses[x][0] < courses[y][0];
        }};
        std::priority_queue<int, vector<int>, decltype(f)> pq(f);
        int day{0}, ans{0};
        for (int i = 0;i < courses.size();++i) {
            if (day + courses[i][0] <= courses[i][1]) {
                ++ans;
                day += courses[i][0];
                pq.push(i);
            } else if (!pq.empty() && courses[pq.top()][0] > courses[i][0] && day - courses[pq.top()][0] + courses[i][0] <= courses[i][1]) {
                day = day - courses[pq.top()][0] + courses[i][0];
                pq.pop();
                pq.push(i);
            }
        }
        return ans;
    }
};